http/basic/src/main/java/io/cloudevents/http/HttpMessageFactory.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2018-Present The CloudEvents Authors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cloudevents.http;
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpsExchange;
+import io.cloudevents.core.message.MessageReader;
+import io.cloudevents.core.message.MessageWriter;
+import io.cloudevents.core.message.impl.GenericStructuredMessageReader;
+import io.cloudevents.core.message.impl.MessageUtils;
+import io.cloudevents.core.message.impl.UnknownEncodingMessageReader;
+import io.cloudevents.http.impl.CloudEventsHeaders;
+import io.cloudevents.http.impl.HttpMessageReader;
+import io.cloudevents.http.impl.HttpMessageWriter;
+
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URLConnection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import static io.cloudevents.http.impl.CloudEventsHeaders.CONTENT_TYPE;
+
+/**
+ * This class provides a collection of methods to create {@link io.cloudevents.core.message.MessageReader}
+ * and {@link io.cloudevents.core.message.MessageWriter} for various HTTP APIs.
+ */
+public final class HttpMessageFactory {
+
+    private HttpMessageFactory() {}
+
+    /**
+     * Creates a new {@link MessageReader} that can read both structured and binary messages from a HTTP response (client) or request (server).
+     *
+     * <pre>
+     * Example of usage with <a href="https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html">HttpServletRequest</a>:
+     * {@code
+     * Consumer<BiConsumer<String,String>> forEachHeader = processHeader -> {
+     *     Enumeration<String> headerNames = httpServletRequest.getHeaderNames();
+     *     while (headerNames.hasMoreElements()) {
+     *         String name = headerNames.nextElement();
+     *         processHeader.accept(name, httpServletRequest.getHeader(name));
+     *
+     *     }
+     * };
+     * byte[] body;
```

Comment:
I think you could use `InputStream.readAllBytes()` in this example. People who are stuck on Java 8 can figure out how to achieve the same thing. I also don't think there's any harm in ignoring the `Content-Length`.

Reply:
I think that with known size the byte array can be allocated straight away. With unknown size there could be reallocation. But who knows what optimizations are under the hood.

Reply:
And yes, it's just an example, but then again some people are going to just copypaste this.